src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -196,30 +207,32 @@ public MemoryPoolIterator ConsumingStart()
                 else
                 {
                     returnStart = _head;
-                    returnEnd = null;
                     _head = null;
                     _tail = null;
                 }
 
-                ReturnBlocks(returnStart, returnEnd);
-
-                if (!_consuming)
```

Comment:
_consuming can be checked and modified outside of the lock. The consumer shouldn't be making concurrent calls anyway.